Managed and Unmanaged subclasses of RealmInteger; partial implementation of the proxy

bmeike opened the pull request:
Separate the base type RealmInteger into its Managed and Unmanaged subtypes.

Create begin the implementation of the proxy and fix the corresponding UTs.

bmeike commented:
I have uncommented the UTs, so this will not pass CI any more.  I would, none the less, like to get review.

beeender commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -65,39 +99,7 @@ public void generate() throws IOException, UnsupportedOperationException {
-        imports.add("io.realm.internal.Table");
-        imports.add("io.realm.internal.OsObject");
-        imports.add("io.realm.internal.SharedRealm");
-        if (!metadata.getBacklinkFields().isEmpty()) {
```
How do you deal with the lint warning may happen to users if the object doesn't have backlink fields?

cmelchior replied:
This check seems to be gone from the changed imports, so something is definitely not right here?

bmeike replied:
I see.  So the generated code gets compiled and should not generate any lint warnings.

What would you think of a @SupressWarnings("all") at the head of each class?  If there *are* warnings, there is nothing that the user can do about them.  While I think that it would make a lot of sense for us to remove that @SupressWarniings in debug versions, so that *we* can make sure our code is good, perhaps it would be smartest to turn them off in release code.

cmelchior replied:
I'm fine with adding @SuppressWarnings("all") at the top of the file if it works, but e.g ErrorProne will catch warnings in the generated files, so it is requirement that we don't have any. There was talk about adding @Generated annotation as well at some point, but afaik that isn't available on Android.

bmeike replied:
Ok.  Got it.  Will revert.  That's a bummer.  A few things like this could speed up our processing quite a bit.

beeender commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -16,31 +16,32 @@
      */
-    public RealmInteger(long value) {
-        this.value = value;
+    public static RealmInteger valueOf(long value) {
```
Since this can only return a unmanaged `RealmInteger`, I don't think it is a useful API?

cmelchior replied:
It would be needed for e.g creating initial values in model classes, just like we have an unmanaged RealmList:

Example:
```
public class Party extends RealmModel {
  private RealmInteger guests = RealmInteger.valueOf(1); 
}
```

beeender commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -16,31 +16,32 @@
      */
-    public RealmInteger(String value) {
-        this.value = Long.parseLong(value);
+    public static RealmInteger valueOf(String value) {
```
same

beeender commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -50,111 +51,70 @@ public RealmInteger(String value) {
-    public int hashCode() {
-        return (int) (value ^ (value >>> 32));
+    public final int hashCode() {
+        long thisValue = longValue();
```
when the managed `RealmInteger` becomes valid, `longValue()` will throw an exception. in this case, the `hashCode()` will throw. But it doesn't sounds like a user expected behaviour? Same for `equals()`. `equals()` would be simpler i think, any invalid `RealmInteger` should not equal to anything.

bmeike replied:
Why will `longValue` throw?

beeender replied:
If it doesn't throw, what should `longValue()` return when teh `RealmInteger` is invalid? (corresponding Realm is closed.)

bmeike replied:
Ahh!!!!  You meant "invalid"!!  Got it.
I think the semantics of `equals` and `hashCode` are correct.  I totally take your point, though.  So, if a managed `RealmInteger` is `isValid() == false`, `equals` should return `false`?  I'll buy that.  What do you think `hashCode` should do?

bmeike replied:
This one will too.  Is that ok with everyone?

bmeike replied:
This one will too.  `longValue()` will, as @beeender says, throw.  Is that what you all want to happen?

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java:
```diff
@@ -15,7 +13,9 @@
 import io.realm.internal.Row;
 import io.realm.internal.SharedRealm;
 import io.realm.internal.Table;
+import io.realm.internal.UncheckedRow;
```
This looks like a mistake that might trigger unused import warnings by static analysis tools.

bmeike replied:
Got it.  Please see comment above.  I suggest @SuppressWarnings("all") in the header of these classes.  Just suggesting.  It seems like a reasonable idea to me but I bow before the opinion of the team.

cmelchior commented on realm/realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java:
```diff
@@ -15,7 +13,9 @@
 import io.realm.internal.Table;
+import io.realm.internal.UncheckedRow;
 import io.realm.internal.android.JsonUtils;
+import io.realm.internal.counters.ManagedRealmInteger;
```
Same

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java:
```diff
@@ -82,8 +83,38 @@ public void basic() {
     @Test
-    public void getters() {
-        RealmInteger ri = new RealmInteger(0x5555444433332211L);
+    public void unmanaged_getters() {
```
To follow the style used elsewhere, this should be reversed so `getters_unmanaged` (<functionUndertest_useCase>). Same for the other tests

bmeike replied:
mmm...  sot the finest grained part comes first?  Like writing 01-05-2017?  That means that the names do not sort into categories.

cmelchior replied:
It is sorted into categories, by the function under test? In this case you get all `getters_<x>`, `setters_<x>` and so on instead of `unmanaged_<x>` and `managed_<x>`. You approach is isn't wrong as such and kinda makes sense since you have more heavy-weight unit tests than normal, but it just break the style of all other unit tests.

bmeike replied:
Ok.  On it.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java:
```diff
@@ -233,54 +363,125 @@ public void copyFromIsNotLive() {
+    }
+
+    /**
+     * Semantics definition from #4266.
```
Link directly to the comment in the issue or better yet, move the description here. Same for the others

bmeike replied:
mmmm... that is not my intention at all.  This test *is* that one comment.  My intention is to link to the entire discussion that surrounds the provisionally agreed semantics that the test represents.  We didn't arrive at these semantics out of thin air.  In fact, there are still reservations about what to do.  My intention is that I connect this test to the entire discussion that created it.  I will add the complete link.  How's that sound?

cmelchior replied:
Complete link is also good, as long as it easy to reach the discussion. I think adding them in {@link x} tags makes it possible to jump directly to it?

bmeike replied:
Yeah!  Using `@see`!

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java:
```diff
@@ -233,54 +363,125 @@ public void copyFromIsNotLive() {
+     */
+    @Test
+    @SuppressWarnings("ReferenceEquality")
+    public void mixed_semantics() {
```
`mixedManagedUnmanagedRealmIntegers_semantics` would explain this test better

bmeike replied:
ok.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java:
```diff
@@ -233,54 +363,125 @@ public void copyFromIsNotLive() {
+        realm.commitTransaction();
+
+        Counters obj2 = new Counters();
+        obj1.columnRealmInteger = RealmInteger.valueOf(30);
```
This should crash? It should never be allowed to modify a managed object outside a transaction

bmeike replied:
Good catch.  That's just a typo.  Should be `obj2`.

cmelchior commented on realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java:
```diff
@@ -233,54 +363,125 @@ public void copyFromIsNotLive() {
-            managedRI.decrement(1);
-            fail("Decrementing a managed RealmInteger outside a transaction should fail");
+            obj2.columnRealmInteger = obj1.columnRealmInteger;
+            fail("Assigning a managed RealmInteger into an unmanaged object should fail");
```
There is something about allowing https://github.com/realm/realm-java/pull/4748/files#diff-41aa371f6cf03f255bad1cd2a85bf66aR415, but disallowing this that makes me really uneasy, but I guess there isn't a way around this without other semantics getting weird instead.

bmeike replied:
Yeah.  Like I said, I would jump at a consistent meaning for this.  I can't come up with one.  If anyone has suggestions, I'd be all over it.

cmelchior commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -50,111 +51,70 @@ public RealmInteger(String value) {
     /**
-     * {@inheritDoc}
+     * RealmIntegers compare strictly by their values.
+     * Final to ensure contract over all subclasses
```
Do we want this description in the public Java doc?

bmeike replied:
Yes, I think we do.  My intention is that it be obvious that this comparison is, exactly, Long comparison.  Does that sound right to you?

cmelchior replied:
It was more the `final to ensure contract` comment. It sounds like an internal comment since marking it final kinda already means users cannot override it?

bmeike replied:
Right.  Equals needs to be reflexive and cannot, in general, be overridden in other than sealed classes/algebraic types.  I'll move the comment.

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/counters/ManagedRealmInteger.java:
```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.internal.counters;
```
I would prefer not to call this package `counters` since this is strictly speaking not just a counter. `io.realm.internal.datatypes` perhaps?

bmeike replied:
I would prefer that if someone creates another datatype, they put it into another package.  I would prefer that this package not become a catch-all for random datatypes.  Would you go for `io.realm.internal.datatypes.counters`?

cmelchior replied:
I'm not a terrible fan of `counters` since this is more than a counter but `integer` is really not better, what about `io.realm.internal.datatypes.realminteger`?

bmeike replied:
DOne!

cmelchior commented on realm/realm-library/src/main/java/io/realm/internal/counters/ManagedRealmInteger.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * Sets the {@code UnmanagedRealmInteger} value.
```
Copy past error here and other places. This is the ManagedRealmInteger

bmeike replied:
Akk!  Thanks!

bmeike commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java:
```diff
@@ -243,6 +243,18 @@ public static void emitFillRealmListFromStream(String interfaceName, String gett
     //@formatter:on
 
+    // @formatter:off
+    public static void emitFillRealmIntegerFromStream(String interfaceName, String getter, String setter, String fieldName, JavaWriter writer) throws IOException {
```
I don't think so.  The semantics are pretty different.  If it turns out to be very similar, I will re-combine things.

Zhuinden commented on realm-transformer/src/main/groovy/io/realm/transformer/BytecodeModifier.groovy:
```diff
@@ -152,11 +152,12 @@ class BytecodeModifier {
-//                    'throw new IllegalStateException("Cannot assign a managed RealmInteger to an unmanaged object")');
-//        }
+        logger.debug "##### SETTER: for ${methodName}: " + field.getType().getName()
+        if ("io.realm.RealmInteger".equals(field.getType().getName())) {
```
Proguard will eat this unless you do something about it.

bmeike replied:
I thought so too!   I think, though, that since this happens at compile time, way before Proguard gets its evil little hands on things, it will work fine.

bmeike replied:
I originally thought that too!

Zhuinden replied:
ah then it's all good :smile:

bmeike commented:
I believe I have addressed all of the comments and that this is ready for the feature branch.  Verified that the Transformer is emitting the correct new synthetic accessor.  Javassist is pretty cool...

zaki50 commented on realm-transformer/src/main/groovy/io/realm/transformer/BytecodeModifier.groovy:
```diff
@@ -147,4 +149,15 @@ class BytecodeModifier {
+        if ("io.realm.RealmInteger".equals(field.getType().getName())) {
+            setter.insertBefore(
+                    'if (!($1 instanceof io.realm.internal.datatypes.realminteger.UnmanagedRealmInteger))' +
+                    'throw new IllegalStateException("Cannot assign a managed RealmInteger to an unmanaged object");');
```
This setter is only for unmanaged object, isn't it?

If so, I don't think that we need to reject `UnmanagedRealmInteger`.

At least all existing setters are only for unmanaged objects since those are overridden by setters defined in the corresponding proxy class.

zaki50 replied:
Ah, I overlooked `!`.

But why do we need to reject managed `RealmInteger`?
I don't think we should interfere unmanaged object. It's normal Java Object.
We should allow everything Java is allowing.

bmeike replied:
I have been convinced on this.  As ugly as it seems, allowing assignment of a managed RealmInteger, to a field in an unmanaged RealmObject, is, as @zaki50 points out, more consistent with the behaviour of the rest of our API.

beeender commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -16,31 +16,37 @@
 @Beta
-public final class RealmInteger extends Number implements Comparable<RealmInteger> {
-    private long value;
+public abstract class RealmInteger extends Number implements Comparable<RealmInteger>, ManagableObject {
```
This need `@Keep` annotation to avoid being obfuscated I think?

beeender commented on realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java:
```diff
@@ -284,6 +288,16 @@ protected final void copy(ColumnInfo rawSrc, ColumnInfo rawDst) {
+    }
+
+    @Override
+    public void realmSet$columnCounter(RealmInteger value) {
```
This feels wrong like @zaki50 said. But maybe you plan to change it in a following PR?

bmeike replied:
Open to suggestion, but it seems right to me.   By all means propose an alternative.

beeender commented on realm/realm-annotations-processor/src/test/resources/io/realm/BooleansRealmProxy.java:
```diff
@@ -29,6 +27,7 @@
 import org.json.JSONException;
 import org.json.JSONObject;
 
+@SuppressWarnings("all")
```
uh, why?

bmeike replied:
Removed.

cmelchior commented:
A good point brought up by @nirinchev. How do we treat `@LinkingObjects` and `RealmList`. Unfortunately, we treat them differently, but it is probably an argument for not introducing a _3rd_ way for `RealmInteger`.

zaki50 commented on realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java:
```diff
@@ -284,6 +288,16 @@ protected final void copy(ColumnInfo rawSrc, ColumnInfo rawDst) {
+
+    @Override
+    public void realmSet$columnCounter(RealmInteger value) {
+        throw new IllegalStateException("Cannot assign to a managed RealmInteger");
```
If the `value` is `null`, we should nullify the field.

And I think that it's better just to set `value.longValue()` if the `value` is not `null` in order to provide consistent behavior because we are doing so in `insertOrUpdate()` https://github.com/realm/realm-java/pull/4748/files#diff-9bbf8752325de6eed4a407ae41857948R1069
And under current implementation, `update()` always throws the exception. https://github.com/realm/realm-java/pull/4748/files#diff-9bbf8752325de6eed4a407ae41857948R1230

bmeike replied:
We seem to be having trouble understanding progressive implementation here.  This is a feature branch.  Each commit moves it towards completion.  There is no intention that this is the complete implementation.

bmeike replied:
I totally give up on this issue.  Done.  Will implement as advised.

zaki50 commented:
I reviewed all changes. Apart from above comment, I put :+1:

cmelchior commented:
@bmeike Crap, I forgot to submit my review points. Feel free to tackle them later

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmInteger.java:
```diff
@@ -50,111 +56,67 @@ public RealmInteger(String value) {
         if (!(o instanceof RealmInteger)) { return false; }
-        RealmInteger other = (RealmInteger) o;
-        return other.value == value;
+        return longValue() == ((RealmInteger) o).longValue();
```
I don't have a lot of experience with these kinds of counters.  I might have the semantics of `.equals` wrong.  My take was that, between `==` and `.equals`, with the semantics I've currently chosen, you get as much information as possible.

bmeike commented on realm/realm-library/src/main/java/io/realm/internal/datatypes/realminteger/UnmanagedRealmInteger.java:
```diff
@@ -0,0 +1,121 @@
+
+
+/**
+ * TODO: document the behaviour of this object, as soon as the implementation defines it.
```
?  Or make it thread safe, I suppose...  I think that is probably the wrong thing to do.  Your average, garden-variety Pojo is not thread safe...

bmeike commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -2073,6 +2115,9 @@ private static int countModelOrListFields(Collection<VariableElement> fields) {
             return type;
         }
+        if (Utils.isRealmInteger(field)) {
+            return Constants.RealmFieldType.COUNTER;
```
Yes. definitely.

bmeike commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java:
```diff
@@ -419,7 +428,7 @@ private void categorizeRequiredField(Element element, VariableElement variableEl
             Utils.error(String.format(
                     "@Required annotation is unnecessary for primitive field \"%s\".", element));
-        } else if (Utils.isRealmList(variableElement) || Utils.isRealmModel(variableElement)) {
+        } else if (Utils.isRealmInteger(variableElement) || Utils.isRealmModel(variableElement) || Utils.isRealmList(variableElement)) {
```
I believe they should *always* be required and not nullable.  I was just having this discussion with @zaki50 .  I don't understand, at all, what it would mean for a counter to be null.

zaki50 replied:
I guess there is a case that only a few objects in a class have counter.

bmeike replied:
NVM.  Will allow nullable RealmIntegers.  NP.

bmeike commented on realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java:
```diff
@@ -369,6 +379,19 @@ public void emit(JavaWriter writer) throws IOException {
+
+        writer.emitAnnotation("Override");
+        writer.beginMethod("void", metadata.getInternalSetter(fieldName), EnumSet.of(Modifier.PUBLIC), fieldTypeCanonicalName, "value")
+                .emitStatement("throw new IllegalStateException(\"Cannot assign to a managed RealmInteger\")")
```
Not that I recall.  The conclusion over there was that assigning a managed object to an unmanaged object was allowed.

bmeike replied:
I take it back.  Will implement.
